---
title: "Allelic Frequencies"
author: "Makar Dorohuntsev"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE,fig.width = 12,fig.height = 5)
```

### 1. Create genind from VCF

Read previously filtered VCF file to R and select most frequent SNP per marker. Next create genind with set populations and group them using seppop() to make list of geninds by populations.

```{r}
library(vcfR)
library(pegas)
library(dplyr)
# load file with population names 
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/pop name.RData")
# snps - Large vcfR data file
load("~/Desktop/Data do projektu/Data/id by mafs.RData")
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/SNP names.RData")
snps <- vcfR::read.vcfR("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/final.vcf", convertNA  = TRUE)
# read row snps from vcf file, transpose and create loci 
row_vcf <- as.loci(t(vcfR::extract.gt(snps, 
                             element = "GT",
                             IDtoRowNames  = T,
                             as.numeric = F,
                             convertNA = F,
                             return.alleles = T)))
#filter loci 
filtred_loci <- row_vcf %>% select(c(unlist(filtred_id)))
colnames(filtred_loci) <- c(unlist(filtred_id))
# convert loci to genind
test_genind <- loci2genind(filtred_loci)
# add population data to genind
test_genind@pop <- as.factor(c(a))
# check percent of missing data 
summary(test_genind)[5]
```

### 2. Using pop.freq() calculate allele frequency for every population

Now create a list of 102 elements, one for each locus. Every element is a list of 11 populations with calculated allelic frequencies for each allele

```{r}
library(hierfstat)
# load marker names
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/SNP names.RData")
# create frequency data
freq_count <- pop.freq(test_genind)
# add marker names to data
names(freq_count) <- filtred_id
```

### 3. For MAF creation may be used maf() command and snps data set created previously

Plot with a frequency distribution between markers in MAF and blue line for mean frequency value.

Note: Every frequency value is higher than 0 but can be very close to it.

```{r}
library(ggplot2)
load("~/Desktop/Data do projektu/Data/id list.RData")
# Data frame creation
Maf2 <- maf(snps,element = 2)
Maf2 <- data.frame(Maf2)
Maf2$`SNP groups` <- c(id_list)
Maf2$`SNP names` <- c(SNP_names)
# Grouping by name and frequency 
grouped_maf<- Maf2 %>%
  group_by(Maf2$`SNP groups`) 
freq_maf <- grouped_maf %>%
  filter(Frequency == max(Frequency))
cuted_test <- freq_maf %>%
  distinct(`SNP groups`, .keep_all = TRUE)
# Visualizing
cuted_test$`Maf2$\`SNP groups\`` <- NULL
cuted_test$`SNP groups` <- NULL

ggplot(cuted_test, aes(x=`Frequency`)) + 
  geom_histogram( colour="black", fill="white",bins = 50)+
  labs(x="Frequency of useful markers (with frequency > 0)",y="Number of markers")+ 
  geom_vline(aes(xintercept=mean(`Frequency`)),
            color="blue", linetype="dashed", size=1)
```
