---
title: "PCA"
author: "Makar Dorohuntsev"
date: "`r Sys.Date()`"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE,fig.width = 11,fig.height = 10)
```

### For PCA calculation were used two guide links:

1\. <https://rstudio-pubs-static.s3.amazonaws.com/975097_5afd0f5ab3fe4d76a93faccff2fda3c2.html>

2\. <https://rstudio-pubs-static.s3.amazonaws.com/981490_b3488a5aa5b94255bab751ad01be4ffd.html>

All credits go to the author of this links.

#### First step is creation of data frame from VCF file

```{r}

library(vcfR)
library(vegan)
#create data with snps
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/pop name.RData")
# snps - Large vcfR data file
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/SNP names.RData")
load("~/Desktop/Data do projektu/Data/id by mafs.RData")
snps <- vcfR::read.vcfR("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/final.vcf", convertNA  = TRUE)
# load population id
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/population id.RData")
#make it numeric matrix where AA get 0, aa get 2 and Aa get 1 values
snps_num <- vcfR::extract.gt(snps, 
                             element = "GT",
                             IDtoRowNames  = F,
                             as.numeric = T,
                             convertNA = T,
                             return.alleles = F)
#Transposition
snps_num_t <- t(snps_num) 
#Make dataframe
snps_num_df <- data.frame(snps_num_t)

colnames(snps_num_df) <- c(SNP_names)
snps_num_df <- snps_num_df[,c(unlist(filtred_id))]
```

#### Next is clearing data frame from NAs to be able to work with data

```{r}

find_NAs <- function(x){
  NAs_TF <- is.na(x)
  i_NA <- which(NAs_TF == TRUE)
  N_NA <- length(i_NA)
  
  #possibility to show removed NAs : cat("Results:",N_NA, "NAs present\n.")
  return(i_NA)
}

N_rows <- nrow(snps_num_t)

# N_NA
# vector to hold output (number of NAs)
N_NA   <- rep(x = 0, times = N_rows)

# N_SNPs
# total number of columns (SNPs)
N_SNPs <- ncol(snps_num_t)

# the for() loop
for(i in 1:N_rows){
  
  # for each row, find the location of
  ## NAs with snps_num_t()
  i_NA <- find_NAs(snps_num_t[i,]) 
  
  # then determine how many NAs
  ## with length()
  N_NA_i <- length(i_NA)
  
  # then save the output to 
  ## our storage vector
  N_NA[i] <- N_NA_i
}

# Delete all rows with more than 40% NA
cutoff40 <- N_SNPs*0.4

percent_NA <- N_NA/N_SNPs*100

# Call which() on percent_NA
i_NA_40percent <- which(percent_NA > 40) 
# Remove rows with more than 40% NA
snps_num_t02 <- snps_num_t[-i_NA_40percent, ]
```

#### Now clean invariant columns which we cant scale

```{r}
invar_omit <- function(x){
  cat("Dataframe of dim",dim(x), "processed...\n")
  sds <- apply(x, 2, sd, na.rm = TRUE)
  i_var0 <- which(sds == 0)
  
  
  cat(length(i_var0),"columns removed\n")
  
  if(length(i_var0) > 0){
    x <- x[, -i_var0]
  }
  
  ## add return()  with x in it
  return(x)                      
}


snps_no_invar <- invar_omit(snps_num_t02) 
```

#### Use mean imputation to delete any left NAs

```{r}
snps_noNAs <- snps_no_invar

N_col <- ncol(snps_no_invar)
for(i in 1:N_col){
  
  # get the current column
  column_i <- snps_noNAs[, i]
  
  # get the mean of the current column
  mean_i <- mean(column_i, na.rm = TRUE)
  
  # get the NAs in the current column
  NAs_i <- which(is.na(column_i))
  
  # record the number of NAs
  N_NAs <- length(NAs_i)
  
  # replace the NAs in the current column
  column_i[NAs_i] <- mean_i
  
  # replace the original column with the
  ## updated columns
  snps_noNAs[, i] <- column_i
  
}
```

#### Perform PCA on scaled data

```{r}

SNPs_cleaned <- snps_noNAs

SNPs_scaled <- scale(SNPs_cleaned)

pca_scaled <- prcomp(SNPs_scaled)
```

#### Next extract PCA scores

```{r}
pca_scores <- vegan::scores(pca_scaled)
new_pca <- data.frame(pca_scores)
#pop_id was set to jesien/wiosna dependent on -j- or -w- part of marker name
#with two grups we can make plot organised by shape and color
new_pca <- data.frame(pop_id,
                          pca_scores)
# remove irrelevant outliners
temporal <- new_pca[-c(94,166),]
new_pca <- temporal
```

#### After checking outlines on PCA plot two most distant individuals were considered to be errant and deleted creating new_pca data frame

```{r}
load("~/Library/CloudStorage/Dropbox/Salamandra project/sorted_BAM/pca data.RData")
```

#### With PCA scores we can create plots and see results

Multiplot() function definition credits :

[http://www.cookbook-r.com/Graphs/Multiple_graphs_on_one_page\_(ggplot2)/](http://www.cookbook-r.com/Graphs/Multiple_graphs_on_one_page_(ggplot2)/){.uri}

```{r include=FALSE}
  multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
    library(grid)
    
    # Make a list from the ... arguments and plotlist
    plots <- c(list(...), plotlist)
    
    numPlots = length(plots)
    
    # If layout is NULL, then use 'cols' to determine layout
    if (is.null(layout)) {
      # Make the panel
      # ncol: Number of columns of plots
      # nrow: Number of rows needed, calculated from # of cols
      layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                       ncol = cols, nrow = ceiling(numPlots/cols))
    }
    
    if (numPlots==1) {
      print(plots[[1]])
      
    } else {
      # Set up the page
      grid.newpage()
      pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
      
      # Make each plot, in the correct location
      for (i in 1:numPlots) {
        # Get the i,j matrix positions of the regions that contain this subplot
        matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
        
        print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                        layout.pos.col = matchidx$col))
      }
    }
  }
```

```{r}
# simply changing PCX and PCY we can get every plot we need 
plotPC1vPC2 <- ggpubr::ggscatter(data = new_pca,
                  y = "PC2",
                  x = "PC1",
                  color = "pop_id",
                  size = 4,
                  xlab = "PC1",
                  ylab = "PC2",
                  main = "Species Variation")
plotPC2vPC3 <- ggpubr::ggscatter(data = new_pca,
                  y = "PC3",
                  x = "PC2",
                  color = "pop_id",
                  size = 4,
                  xlab = "PC2",
                  ylab = "PC3",
                  main = "Species Variation")
plotPC3vPC4 <- ggpubr::ggscatter(data = new_pca,
                  y = "PC4",
                  x = "PC3",
                  color = "pop_id",
                  size = 4,
                  xlab = "PC3",
                  ylab = "PC4",
                  main = "Species Variation")

multiplot(plotPC1vPC2,plotPC2vPC3,plotPC3vPC4)
```

Information on the plots allows to assume, that there are no specific outlining groups of populations, only some individuals.
